libbpf-tools/offwaketime.bpf.c

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2022 Ze Gao
+
+#include <vmlinux.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "offwaketime.h"
+
+#define PF_KTHREAD 0x00200000 /* I am a kernel thread */
+#define MAX_ENTRIES 10240
+
+const volatile bool kernel_threads_only = false;
+const volatile bool user_threads_only = false;
+const volatile __u64 max_block_ns = -1;
+const volatile __u64 min_block_ns = 1;
+const volatile pid_t targ_tgid = -1;
+const volatile pid_t targ_pid = -1;
+const volatile long state = -1;
+
+struct {
+        __uint(type, BPF_MAP_TYPE_HASH);
+        __type(key, u32);
+        __type(value, u64);
+        __uint(max_entries, MAX_ENTRIES);
+} start SEC(".maps");
+
+struct {
+        __uint(type, BPF_MAP_TYPE_STACK_TRACE);
+        __uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+struct {
+        __uint(type, BPF_MAP_TYPE_HASH);
+        __type(key, u32);
+        __type(value, struct pkey_t);
+        __uint(max_entries, MAX_ENTRIES);
+} wokeby SEC(".maps");
+
+struct {
+        __uint(type, BPF_MAP_TYPE_HASH);
+        __type(key, struct count_key_t);
+        __type(value, u64);
+        __uint(max_entries, MAX_ENTRIES);
+} count SEC(".maps");
+
+static inline bool allow_record(u32 pid, u32 tgid, u32 flags, long pstate) {
+        if (targ_tgid != -1 && targ_tgid != tgid)
+                return false;
+        if (targ_pid != -1 && targ_pid != pid)
+                return false;
+        if (user_threads_only && flags & PF_KTHREAD)
+                return false;
+        else if (kernel_threads_only && !(flags & PF_KTHREAD))
+                return false;
+        if (state != -1 && pstate != state)
+                return false;
+        return true;
+}
+
+struct task_struct___new {
+        long __state;
+} __attribute__((preserve_access_index));
+
+SEC("kprobe/try_to_wake_up")
+int BPF_KPROBE(try_to_wake_up, struct task_struct *p) {
+        struct pkey_t woke;
+        u32 pid, tgid, flags, pstate;
+        pid = BPF_CORE_READ(p, pid);
+        tgid = BPF_CORE_READ(p, tgid);
+        flags = BPF_CORE_READ(p, flags);
+        struct task_struct___new *p_new = (void *)p;
+        if (bpf_core_field_exists(p_new->__state)) {
+                pstate = BPF_CORE_READ(p_new, __state);
+        } else {
+                pstate = BPF_CORE_READ(p, state);
+        }
```

Comment:
I'm afraid it's used for btf-enabled prog only cuz `return ((struct task_struct *)task)->state;` directly loads state here instead of accessing it in a CO-relocatable way.  call it inside a kprobe bpf prog will not compile. error like this:

```sh

; return t->__state;
29: (61) r1 = *(u32 *)(r9 +0)
R9 invalid mem access 'inv'
processed 30 insns (limit 1000000) max_states_per_insn 0 total_states 1 peak_states 1 mark_read 1
```

So here it solves a different problem.

Reply:
I see, this will be addressed by #3691.

Reply:
or we can modify **get_task_state** in https://github.com/iovisor/bcc/blob/5e3d41e21d936168f807e050ddd932c4553e3469/libbpf-tools/core_fixes.bpf.h#L20-L27  to make it more general?

Reply:
#3691 is landed, you can reuse it now.